internal/resources/slo/resource_slo.go

```diff
@@ -0,0 +1,477 @@
+	"github.com/grafana/terraform-provider-grafana/internal/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func ResourceSlo() *schema.Resource {
+	return &schema.Resource{
+		Description: `
+Resource manages Grafana SLOs. 
+
+* [Official documentation](https://grafana.com/docs/grafana-cloud/slo/)
+* [API documentation](https://grafana.com/docs/grafana-cloud/slo/api/)
+		`,
+		CreateContext: resourceSloCreate,
+		ReadContext:   resourceSloRead,
+		UpdateContext: resourceSloUpdate,
+		DeleteContext: resourceSloDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Name should be a short description of your indicator. Consider names like "API Availability"`,
+			},
+			"description": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Description is a free-text field that can provide more context to an SLO.`,
+			},
+			"query": &schema.Schema{
+				Type:        schema.TypeList,
+				MaxItems:    1,
+				Required:    true,
+				Description: `Query describes the indicator that will be measured against the objective. Freeform Query types are currently supported.`,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: `Query type must be one of: "freeform", "query", "ratio", or "threshold"`,
+							Required:    true,
+						},
+						"freeformquery": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"labels": &schema.Schema{
+				Type:        schema.TypeList,
+				Optional:    true,
+				Description: `Additional labels that will be attached to all metrics generated from the query. These labels are useful for grouping SLOs in dashboard views that you create by hand.`,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"value": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"objectives": &schema.Schema{
+				Type:        schema.TypeList,
+				Required:    true,
+				Description: `Over each rolling time window, the remaining error budget will be calculated, and separate alerts can be generated for each time window based on the SLO burn rate or remaining error budget.`,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"value": &schema.Schema{
+							Type:        schema.TypeFloat,
+							Required:    true,
+							Description: `Value between 0 and 1. If the value of the query is above the objective, the SLO is met.`,
```

Comment:
`ValidateFunc` here

Reply:
Thanks for this! I didn't even know the `ValidateFunc`s were a thing (: